packages/core/core/src/public/BundlerBundleGraph.js

```diff
@@ -0,0 +1,280 @@
+          actions.skipChildren();
+        }
+      }
+    );
+
+    if (connectedFromBundles.length > 0) {
+      for (let bundleNode of connectedFromBundles) {
+        this.#graph._graph.addEdge(bundleNode.id, bundleGroupNode.id, 'bundle');
+      }
+    } else {
+      this.#graph._graph.addEdge(
+        nullthrows(this.#graph._graph.getRootNode()).id,
+        bundleGroupNode.id,
+        'bundle'
+      );
+    }
+
+    return bundleGroup;
+  }
+
+  createBundle(
+    opts:
+      | {|
+          id?: string,
+          entryAsset: IAsset,
+          target: Target,
+          isEntry?: ?boolean,
+          type?: ?string,
+          env?: ?Environment
+        |}
+      | {|
+          id: string,
+          entryAsset?: IAsset,
+          target: Target,
+          isEntry?: ?boolean,
+          type: string,
+          env: Environment
+        |}
+  ): Bundle {
+    let bundleId = 'bundle:' + (opts.id ?? nullthrows(opts.entryAsset?.id));
+    let bundleNode = {
+      type: 'bundle',
+      id: bundleId,
+      value: {
+        id: bundleId,
+        type: opts.type ?? nullthrows(opts.entryAsset).type,
+        env: opts.env ?? nullthrows(opts.entryAsset).env,
+        entryAssetId: opts.entryAsset?.id,
+        filePath: null,
+        isEntry: opts.isEntry,
+        target: opts.target,
+        name: null,
+        stats: {size: 0, time: 0}
+      }
+    };
+
+    this.#graph._graph.addNode(bundleNode);
+    if (opts.entryAsset != null) {
+      this.#graph._graph.addEdge(bundleNode.id, opts.entryAsset.id);
+    }
+
+    return new Bundle(bundleNode.value, this.#graph);
+  }
+
+  addBundleToBundleGroup(bundle: IBundle, bundleGroup: BundleGroup) {
+    let bundleGroupId = getBundleGroupId(bundleGroup);
+    this.#graph._graph.addEdge(bundleGroupId, bundle.id);
+    this.#graph._graph.addEdge(bundleGroupId, bundle.id, 'bundle');
+    for (let entryAsset of bundle.getEntryAssets()) {
+      this.#graph._graph.removeEdge(bundleGroupId, entryAsset.id);
+    }
+  }
+
+  createAssetReference(dependency: Dependency, asset: IAsset): void {
+    return this.#graph.createAssetReference(
+      dependency,
+      assetToInternalAsset(asset)
+    );
+  }
+
+  getDependencyAssets(dependency: Dependency): Array<IAsset> {
+    return this.#graph
+      .getDependencyAssets(dependency)
+      .map(asset => new Asset(asset));
+  }
+
+  traverse<TContext>(
+    visit: GraphVisitor<BundlerBundleGraphTraversable, TContext>
+  ): ?TContext {
+    return this.#graph._graph.filteredTraverse(
+      node => {
+        if (node.type === 'asset') {
+          return {type: 'asset', value: new Asset(node.value)};
+        } else if (node.type === 'dependency') {
+          return {type: 'dependency', value: node.value};
+        }
+      },
+      visit,
+      undefined, // start with root
+      // $FlowFixMe
+      ALL_EDGE_TYPES
+    );
+  }
+}
+
+export class BundlerOptimizeBundleGraph extends BundlerBundleGraph
```

Comment:
Do we need the separation of types between initial bundling and optimizing? If we can combine them I think that might be slightly simpler.

Reply:
These are separated for the same reason that `bundle` and `optimize` are, unfortunately. Because it's cheaper for core not to attach `contains` edges downward every time a bundle is created, and removing them every time a child bundle is created, we only "summarize" the bundles post-`bundle` and the implementation details have to behave differently.

One thing we could do to not leak this as much, and to simplify the burden on plugin authors, is send a bundle graph with an identical interface to both phases, but implement them differently depending on the phase.

Reply:
I guess it's cheaper because you would have to remove edges from existing bundles when a child bundle is created. If you were just connecting the edges after the bundles were made, it would be the same number of traversals.

What if you created the bundles in a post-traversal rather than a pre-traversal? That way, child bundles would have already been created when you create the parents. That way, you could just connect the edges as the bundles are created and never need to remove edges from existing bundles. Maybe I'm missing some edge case though.